### ISSUE TITLE
Bottle update will outdate locally compiled version?

### ISSUE THREAD

**Issue by smarek** [Homebrew/homebrew#46006]
Hello,

something I observed many times so far has been bothering me.
I do not use bottles at all (`ENV['HOMEBREW_BUILD_FROM_SOURCE']=true`)

However from time to time, there is update to a Formula which bumps the version to something like `gobject-introspection 1.46.0_1`, which however is just a bottle update, not a source-code or build-script update, and forces me to update the compiled formula locally. Same thing was happening in past months for example with imagemagick.

I think this is principally wrong, and should be marked if the change is made for bottles only or if there is reason to re-compile the formula when built from source.

Proof:
```diff
> git show 0cbab1b96f341db465986d8a7a23d563b600aa18
commit 0cbab1b96f341db465986d8a7a23d563b600aa18
Author: BrewTestBot <brew-test-bot@googlegroups.com>
Date:   Sat Nov 14 22:40:16 2015 +0000

    gobject-introspection: update 1.46.0_1 bottle.

diff --git a/Library/Formula/gobject-introspection.rb b/Library/Formula/gobject-introspection.rb
index 84af61f..aba1a60 100644
--- a/Library/Formula/gobject-introspection.rb
+++ b/Library/Formula/gobject-introspection.rb
@@ -6,9 +6,9 @@ class GobjectIntrospection < Formula
   revision 1
 
   bottle do
-    sha256 "8cc016da6173e849904b707d9183f77078770bcda5db26838c28808b18b93dc0" => :el_capitan
-    sha256 "55644f76743a7d0ff9536876272ad3de9c40fe13e411fb34ccec0f4d4536c96f" => :yosemite
-    sha256 "1a7789dd8f4f693c5f494cace1a9b557bc8e440c1a562325b8d2242020fe63fb" => :mavericks
+    sha256 "6bae714bf138ddf4af57736d593d5cb4c68b2efa8422e8111fdecab0979786f8" => :el_capitan
+    sha256 "3ad3ce2a65ab5622087cc4a50313767c3ca97402b2c4064904a773cbd355c732" => :yosemite
+    sha256 "497496d62580f4d6a41fa285ee164d2d693139dc41cf5c2891225ee6869bb8c2" => :mavericks
   end
 
   option :universal
```

**Comment by DomT4** [comment]
> However from time to time, there is update to a Formula which bumps the version to something like gobject-introspection 1.46.0_1, which however is just a bottle update, not a source-code or build-script update, and forces me to update the compiled formula locally

Actually, this isn't true in the example you provided. `_1` indicates the formula revision was bumped, which means there were some source changes or something needed fixing, as you can see in the https://github.com/Homebrew/homebrew/commit/0ae2be4e9a7e644e85a532764530f3bdc242dbc2 commit, which yes, was followed by new bottles in https://github.com/Homebrew/homebrew/commit/0cbab1b96f341db465986d8a7a23d563b600aa18.

**Comment by mikemcquaid** [comment]
> I do not use bottles at all (ENV['HOMEBREW_BUILD_FROM_SOURCE']=true)

Why do you do this? Sometimes we do have to bump the `revision` for a bottle update and your use-case is sufficiently niche that I don't think the workarounds are worth it, honestly.

**Comment by smarek** [comment]
@mikemcquaid because I can obviously :-)
@DomT4 well strange is to me, that bottle is changed, but there is no change in formula (like changed link to sources or buildscript configuration)

**Comment by mikemcquaid** [comment]
> @mikemcquaid because I can obviously :-)

Can you elaborate on that?

**Comment by DomT4** [comment]
> (like changed link to sources or buildscript configuration)

There wasn't an upstream update, but we applied a patch to resolve some issues others were seeing in related builds. That patch would have produced a slightly different package, hence the need for new bottles. If you didn't get the [patch commit](https://github.com/Homebrew/homebrew/commit/0ae2be4e9a7e644e85a532764530f3bdc242dbc2) locally, there may be some issue with your local configuration.

**Comment by smarek** [comment]
@mikemcquaid the feature is available, and I use it, is there any reason why I should be simply trusting your buildserver everything's ok? I do also verify the checksums of downloaded source packages and have strict ssl sources usage (and ssl certificate pinning in place), I'm comfortable like that.

@DomT4 oh, sorry, that clears it up, I found the commit down the git log, 

```git
commit 0cbab1b96f341db465986d8a7a23d563b600aa18
Author: BrewTestBot <brew-test-bot@googlegroups.com>
Date:   Sat Nov 14 22:40:16 2015 +0000

    gobject-introspection: update 1.46.0_1 bottle.

commit 0ae2be4e9a7e644e85a532764530f3bdc242dbc2
Author: Tom Schoonjans <Tom.Schoonjans@diamond.ac.uk>
Date:   Sat Nov 14 15:08:43 2015 +0000

    gobject-introspection: fix for non-libtool builds
    
    pulls in upstream committed patch from https://bugzilla.gnome.org/show_bug.cgi?id=625195
    
    Closes #45992.
    
    Signed-off-by: Dominyk Tiller <dominyktiller@gmail.com>
```

Anyway, does it happen that you'd do change in your internal build system (BrewTestBot), and that would bumped the package(s) revision without any impact on package build-process or was it always this reason, I've seen quite many revisions for single version of upstream package? Thank you

**Comment by DomT4** [comment]
Updates marked with `_1` or `_2` or so on are (almost certainly) always revision bumps, which means there was a change noteworthy enough to require people to rebuild. This doesn't need to be an upstream new release; we bump revisions to fix breakage, for important patches, because it needs to be rebuilt to work well against an updated dependency, because it links into the old system OpenSSL, etc.

It is fairly often the case that on major updates other packages have to be recompiled to remain compatible, such as the recent `mysql` update and the pending `icu4c` update. If we updated only the core package there people would hit severe breakage, so we mitigate that by bumping the revision of other formulae that are known to be failing in some way.

Pure bottle-only revisions tend to happen because:
* There's an opportunity to generate a bottle for the latest OS X release which doesn't have one.
* There's an opportunity to replace SHA1-tagged bottles which we've deprecated.
* A runtime issue was discovered in the current bottle but is not important enough to merit forcing everyone to rebuild immediately.
* It was part of a large PR and not pulling bottles on 1 formula inside a 30 formula PR isn't worth the trouble.
